module1/homework/grayscaleImages/compression.cpp

```diff
@@ -0,0 +1,60 @@
+#include "compression.hpp"
+
+#include <algorithm>
+#include <iostream>
+#include <utility>
+
+std::vector<std::pair<uint8_t, uint8_t>> compressGrayscale(
+    const std::array<std::array<uint8_t, width>, height>& bitmap) {
+    std::vector<std::pair<uint8_t, uint8_t>> compressed{};
+    compressed.reserve(height * width);
+    int occurrences = 0;
+    uint8_t color_code = 0;
+    for (const auto& row : bitmap) {
+        color_code = *row.begin();
+        occurrences = 0;
+        for (auto element : row) {
+            if (element == color_code) {
+                ++occurrences;
+            } else {
+                compressed.emplace_back(std::make_pair(color_code, occurrences));
+                color_code = element;
+                occurrences = 1;
+            }
+        }
+        compressed.emplace_back(std::make_pair(color_code, occurrences));
+    }
+    compressed.shrink_to_fit();
+    return compressed;
+}
+
+std::array<std::array<uint8_t, width>, height> decompressGrayscale(
+    const std::vector<std::pair<uint8_t, uint8_t>>& bitmap) {
+    std::array<std::array<uint8_t, width>, height> decompressed{};
+    auto it = bitmap.begin();
+    int counter = 0;
+    for (size_t row = 0; row < height; ++row) {
+        for (size_t column = 0; column < width;) {
+            while (counter++ < it->second) {
+                decompressed[row][column++] = it->first;
+            }
+            ++it;
+            counter = 0;
+        }
+    }
+    return decompressed;
+}
+
+void printMap(const std::array<std::array<uint8_t, width>, height>& bitmap) {
+    for (const auto& row : bitmap) {
+        for (auto column : row) {
+            if (column <= ' ') {
```

Comment:
Could you paste the output here? I had really similar soultion but the "weird question marks (ninja mask)" was printed as " ", which is not correct :)

Reply:
I changed it a bit but there is an output:
![image](https://user-images.githubusercontent.com/26752357/84577323-a8297400-adbb-11ea-868f-fb5409e9506d.png)


Now I am using empty_char which is `'  '`

Reply:
Maybe there was a problem with my terminal but i had really hard time with this one ;( some checking `iscntrl` (first 31 ascii's) helps. Anyway, good for you :)